### PR TITLE
Added Spinlock Example/Test 

### PR DESCRIPTION
**Pull request by gyrovorbis**
Upon constructing a test scenario for @ljsebald's last PR (addition of `spinlock_trylock()`), it was decided that we may as well flesh the test case out into an example for KOS. 

The example is located in `kos/examples/dreamcast/basic/theading/spinlock_test`. It spawns 10 threads which all compete for acquiring a spinlock using both `spinlock_trylock()` and `spinlock_lock()`. It also shows off doing something "useful" while `spinlock_trylock()` fails.

### CODE REVIEW

File: examples/dreamcast/basic/threading/spinlock_test/Makefile [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# KallistiOS ##version##
+#
+# basic/threading/spinlock_test/Makefile
+#
+# (c) 2023 Lawrence Sebald
+# (c) 2023 Falco Girgis
```
Comment (ljsebald) [inline examples/dreamcast/basic/threading/spinlock_test/Makefile:6]:
```suggestion
# Copyright (C) 2023 Falco Girgis
```

File: examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c [hunk 2]
```diff
@@ -0,0 +1,153 @@
+/* KallistiOS ##version##
+
+   spinlock_test.c
+
+   (c) 2023 Lawrence Sebald
+   (c) 2023 Falco Girgis
```
Comment (ljsebald) [inline examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c:6]: Not sure why I'm being credited here...

```suggestion
   Copyright (C) 2023 Falco Girgis
```

File: examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+/* 
+    Main execution path for each thread, demonstrating the following
+    with a spinlock:
+
+        1) Doing something useful while waiting for the lock
+        2) Simply locking normally and waiting for it
+*/
+void *thd(void *v) {
+    unsigned tid = (unsigned)v;
+    unsigned fibres=0, fibn=0;
```
Comment (ljsebald) [inline examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c:61]:
```suggestion
    unsigned fibres = 0, fibn = 0;
```

File: examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    /* Continue calling spinlock_trylock() until eventually 
+       locking succeeds. */
+    while(!spinlock_trylock(&lock)) {
+        print("Thread[%u] still trying the lock!\n", tid);
+        /* We can do something else while we wait! */
+        fibres = fib(fibn++);
+        /* Optionally let the other threads have some time too. */
```
Comment (ljsebald) [inline examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c:69]:
```suggestion
        /* Let the other threads have some time too. */
```

File: examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    kthread_t *threads[thread_count]; 
+    int i, join_error = 0;
+    unsigned fibcount = 0;
+
+    print("Starting Threads\n");
+
+    /* Kick off a number of threads to all compete for our spinlock. */
+    for (i = 0; i < thread_count; i++) {
+        threads[i] = thd_create(0, thd, (void *) i + 1);
+    };
```
Comment (ljsebald) [inline examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c:119]:
```suggestion
        threads[i] = thd_create(0, thd, (void *)(i + 1));
    }
```

File: examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+       a second time. */
+    ++lock_counter; 
+    sleep(1);
+
+    print("Thread[%u] unlocking the lock\n", tid);
+    spinlock_unlock(&lock);
+
+    /* Return how far we got into the Fibonacci sequence
+       while we waited for the lock. */
+    return (void*)(fibn - 1);
```
Comment (ljsebald) [inline examples/dreamcast/basic/threading/spinlock_test/spinlock_test.c:105]:
```suggestion
    return (void *)(fibn - 1);
```

**Review by ljsebald — CHANGES_REQUESTED** [review]

**Comment by gyrovorbis** [comment]
Rebuilt after addressing PR changes, everything looks good.
`
..............
Thread[6] yielding the lock
Thread[6] locking the lock
Thread[6] locked the lock
Thread[6] unlocking the lock
Thread[6] returned.
Thread[7] returned.
Thread[8] trylock succeeded!
Thread[8] calculated the 38th fibonacci number while waiting: 39088169
Thread[10] still trying the lock!
Thread[8] yielding the lock
Thread[8] locking the lock
Thread[8] locked the lock
Thread[8] unlocking the lock
Thread[8] returned.
Thread[9] returned.
Thread[10] trylock succeeded!
Thread[10] calculated the 39th fibonacci number while waiting: 63245986
Thread[10] yielding the lock
Thread[10] locking the lock
Thread[10] locked the lock
Thread[10] unlocking the lock
Thread[10] returned.
Threads finished and calculated 339 fibonacci numbers while they waited!


***** SPINLOCK TEST SUCCESS! *****
`

**Review by ljsebald — APPROVED** [review]

**Pull request merged**